core/src/main/java/com/linecorp/armeria/client/AbstractDnsResolverBuilder.java

```diff
@@ -0,0 +1,558 @@
+import io.netty.resolver.dns.NoopDnsCache;
+import io.netty.resolver.dns.NoopDnsCnameCache;
+import io.netty.util.concurrent.EventExecutor;
+
+/**
+ * A skeletal builder implementation for DNS resolvers.
+ */
+@UnstableApi
+public abstract class AbstractDnsResolverBuilder {
+
+    private static final long DEFAULT_QUERY_TIMEOUT_MILLIS = 5000; // 5 seconds.
+
+    private DnsCache dnsCache = DnsCache.ofDefault();
+    private String cacheSpec = Flags.dnsCacheSpec();
+    private int minTtl = 1;
+    private int maxTtl = Integer.MAX_VALUE;
+    private int negativeTtl;
+    private boolean needsToCreateDnsCache;
+
+    private boolean traceEnabled = true;
+    private long queryTimeoutMillis = DEFAULT_QUERY_TIMEOUT_MILLIS;
+    private long queryTimeoutMillisForEachAttempt = -1;
+
+    private boolean recursionDesired = true;
+    private int maxQueriesPerResolve = -1;
+    private int maxPayloadSize = 4096;
+    private boolean optResourceEnabled = true;
+
+    private HostsFileEntriesResolver hostsFileEntriesResolver = HostsFileEntriesResolver.DEFAULT;
+    private DnsServerAddressStreamProvider serverAddressStreamProvider =
+            DnsServerAddressStreamProviders.platformDefault();
+    @Nullable
+    private DnsQueryLifecycleObserverFactory dnsQueryLifecycleObserverFactory;
+    private boolean dnsQueryMetricsEnabled = true;
+    private List<String> searchDomains = DnsUtil.defaultSearchDomains();
+    private int ndots = DnsUtil.defaultNdots();
+    private boolean decodeIdn = true;
+
+    @Nullable
+    private MeterRegistry meterRegistry;
+
+    /**
+     * Creates a new instance.
+     */
+    protected AbstractDnsResolverBuilder() {}
+
+    /**
+     * Sets if this resolver should generate detailed trace information in exception messages so that
+     * it is easier to understand the cause of resolution failure. This flag is enabled by default.
+     */
+    public AbstractDnsResolverBuilder traceEnabled(boolean traceEnabled) {
+        this.traceEnabled = traceEnabled;
+        return this;
+    }
+
+    /**
+     * Sets the timeout of the DNS query performed by this resolver.
+     * {@code 0} disables the timeout.
+     * If unspecified, {@value #DEFAULT_QUERY_TIMEOUT_MILLIS} ms will be used.
+     */
+    public AbstractDnsResolverBuilder queryTimeout(Duration queryTimeout) {
+        requireNonNull(queryTimeout, "queryTimeout");
+        checkArgument(!queryTimeout.isNegative(), "queryTimeout: %s (expected: >= 0)", queryTimeout);
+        return queryTimeoutMillis(queryTimeout.toMillis());
+    }
+
+    /**
+     * Returns the timeout of the DNS query performed by this resolver in milliseconds.
+     */
+    protected final long queryTimeoutMillis() {
+        return queryTimeoutMillis;
+    }
+
+    /**
+     * Sets the timeout of the DNS query performed by this resolver in milliseconds.
+     * {@code 0} disables the timeout.
+     * If unspecified, {@value #DEFAULT_QUERY_TIMEOUT_MILLIS} ms will be used.
```

Comment:
nit; `DEFAULT_QUERY_TIMEOUT_MILLIS` is a private value. Not sure if javadocs will be generated correctly

Reply:
The value is embedded so it is OK.